wallet/src/de/schildbach/wallet/service/CoinJoinService.kt

```diff
@@ -0,0 +1,323 @@
+package de.schildbach.wallet.service
+
+import de.schildbach.wallet.Constants
+import de.schildbach.wallet.WalletApplication
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import org.bitcoinj.coinjoin.CoinJoinClientManager
+import org.bitcoinj.coinjoin.CoinJoinClientOptions
+import org.bitcoinj.coinjoin.PoolMessage
+import org.bitcoinj.coinjoin.PoolStatus
+import org.bitcoinj.coinjoin.callbacks.RequestDecryptedKey
+import org.bitcoinj.coinjoin.callbacks.RequestKeyParameter
+import org.bitcoinj.coinjoin.listeners.MixingCompleteListener
+import org.bitcoinj.coinjoin.listeners.SessionCompleteListener
+import org.bitcoinj.coinjoin.progress.MixingProgressTracker
+import org.bitcoinj.coinjoin.utils.CoinJoinManager
+import org.bitcoinj.coinjoin.utils.ProTxToOutpoint
+import org.bitcoinj.core.AbstractBlockChain
+import org.bitcoinj.core.Coin
+import org.bitcoinj.core.Context
+import org.bitcoinj.utils.Threading
+import org.bitcoinj.wallet.Wallet
+import org.bitcoinj.wallet.WalletEx
+import org.dash.wallet.common.WalletDataProvider
+import org.dash.wallet.common.data.NetworkStatus
+import org.dash.wallet.common.services.BlockchainStateProvider
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.util.concurrent.Executors
+import javax.inject.Inject
+
+interface CoinJoinService {
+    fun needsToMix(amount: Coin): Boolean
+    fun configureMixing(
+        amount: Coin,
+        requestKeyParameter: RequestKeyParameter,
+        requestDecryptedKey: RequestDecryptedKey
+    )
+
+    suspend fun prepareAndStartMixing()
+    fun startMixing(): Boolean
+    fun stopMixing()
+    fun getMixingStatus(): MixingStatus
+    suspend fun waitForMixing()
+    suspend fun waitForMixingWithException()
+}
```

Comment:
Some of these methods will be removed as the main is to have the implementation manage starting/stopping.

Some of these methods are leftover from previous iterations.